test/unit/query_expr_compiler/test_rewrite_rules.py

```diff
@@ -0,0 +1,520 @@
+"""Tests for rewrite rules."""
+from dataclasses import replace
+from typing import Optional, Union
+
+import pytest
+from tmlt.core.domains.spark_domains import (
+    SparkColumnDescriptor,
+    SparkFloatColumnDescriptor,
+    SparkIntegerColumnDescriptor,
+)
+from tmlt.core.measurements.aggregations import NoiseMechanism
+from tmlt.core.measures import ApproxDP, PureDP, RhoZCDP
+
+from tmlt.analytics import KeySet
+from tmlt.analytics._catalog import Catalog
+from tmlt.analytics._query_expr import (
+    AverageMechanism,
+    CountDistinctMechanism,
+    CountMechanism,
+    GroupByBoundedAverage,
+    GroupByBoundedSTDEV,
+    GroupByBoundedSum,
+    GroupByBoundedVariance,
+    GroupByCount,
+    GroupByCountDistinct,
+    PrivateSource,
+    QueryExpr,
+    StdevMechanism,
+    SumMechanism,
+    SuppressAggregates,
+    VarianceMechanism,
+)
+from tmlt.analytics._query_expr_compiler._rewrite_rules import CompilationInfo, rewrite
+from tmlt.analytics._schema import ColumnDescriptor, ColumnType
+
+# SPDX-License-Identifier: Apache-2.0
+# Copyright Tumult Labs 2025
+
+
+@pytest.fixture(name="test_data", scope="class")
+def setup_test(request):
+    """Setup tests."""
+    catalog = Catalog()
+    catalog.add_private_table(
+        "private",
+        {
+            "A": ColumnDescriptor(ColumnType.VARCHAR),
+            "B": ColumnDescriptor(ColumnType.INTEGER),
+            "X": ColumnDescriptor(ColumnType.DECIMAL),
+        },
+    )
+    request.cls.catalog = catalog
+
+
+BASE_EXPR = PrivateSource("private")
+
+
+@pytest.mark.usefixtures("test_data")
+class TestRewriteRules:
+    """Tests for rewrite rules."""
+
+    catalog: Catalog
+    base_expr: QueryExpr
```

Comment:
Grouping tests into classes is not idiomatic for pytest. While we do have a lot of tests structured like this, that's because we weren't using pytest when we wrote most of these tests. I'd prefer to write more normal (for pytest) tests going forward, and I don't see a compelling reason to do otherwise here.

The way to do that is to pull all of these tests out of a class, use the `test_data` fixture on each one, and rewrite it to return a catalog rather than setting one as a class variable. See, e.g., [here](https://github.com/opendp/tumult-analytics/blob/39ba1aeba72d8105542b90011d6442ea9b73f884/test/unit/keysets/test_keyset.py#L816) for an example.